### PR TITLE
Add explicit MSI

### PR DESCRIPTION
**Pull request by lmazuel**
FYI @yugangw-msft

### CODE REVIEW

File: msrestazure/azure_active_directory.py [hunk 1]
```diff
@@ -559,11 +559,19 @@ def signed_session(self):
         return session
 
 
-def get_msi_token(resource, port=50342):
+def get_msi_token(resource, port=50342, client_id=None):
```
Comment (yugangw-msft) [inline msrestazure/azure_active_directory.py]: FYI, msi's resource id and object id are also supported

Comment (lmazuel) [reply msrestazure/azure_active_directory.py]: I saw your email about this naming. Will wait for official suggestion before merging.

File: msrestazure/azure_active_directory.py [hunk 2]
```diff
@@ -632,10 +644,18 @@ def get_msi_token_webapp(resource):
     return token_entry['token_type'], token_entry['access_token'], token_entry
 
 
+def _is_app_service():
+    # Might be discussed if we think it's not robust enough
+    return 'APPSETTING_WEBSITE_SITE_NAME' in os.environ
```
Comment (yugangw-msft) [inline msrestazure/azure_active_directory.py]: I am not familiar with this env variable, but I know we have `MSI_ENDPOINT` and `MSI_SECRET` to leverage

File: msrestazure/azure_active_directory.py [hunk 3]
```diff
@@ -559,11 +559,23 @@ def signed_session(self):
...
 
-def get_msi_token(resource, port=50342):
+def get_msi_token(resource, port=50342, msi_conf=None):
+    """Get MSI token from inside a VM/VMSS.
+
+    If msi_conf is used, must be a dict of one key in ["client_id", "object_id", "msi_res_id"]
+
+    :param str resource: The resource where the token would be use.
+    :param int port: The port is not the default 50342 is used.
+    :param dict[str, str] msi_conf: msi_conf if User Assigned (if not specified, assume System Assigned)
```
Comment (yugangw-msft) [inline msrestazure/azure_active_directory.py]: Suggest wording of `if to request a token through a user assigned identity`. But up to you

File: msrestazure/azure_active_directory.py [hunk 4]
```diff
@@ -647,15 +667,16 @@ def __init__(self, port=50342, **kwargs):
         super(MSIAuthentication, self).__init__(None)
 
         self.port = port
+        self.msi_conf = {k:v for k,v in kwargs.items() if k in ["client_id", "object_id", "msi_res_id"]}
```
Comment (yugangw-msft) [inline msrestazure/azure_active_directory.py]: still debating a bit, but to get interface clear, we should throw when user assigned identity info is supplied for a  webapp

File: msrestazure/azure_active_directory.py (hunk 2 above)
Comment (lmazuel) [reply msrestazure/azure_active_directory.py]: When I did the MSI sample of app service, I realized that I need first to detect if I'm on a WebApp. For that, MSI_ENDPOINT is not good (if MSI is not activated).
APPSETTING_WEBSITE_SITE_NAME is always there, it's a WebApp requirement. This makes possible to detect: if it's a WebApp, no MSI_ENDPOINT is not there, tell the customer.

File: msrestazure/azure_active_directory.py (hunk 4 above)
Comment (lmazuel) [reply msrestazure/azure_active_directory.py]: Let's do that for now, but I won't be surprised if this comes to WebApp at some point...

Comment (lmazuel) [reply msrestazure/azure_active_directory.py]: Done (see last commit)

File: msrestazure/azure_active_directory.py (hunk 2 above)
Comment (yugangw-msft) [reply msrestazure/azure_active_directory.py]: Got it. I don't know whether `APPSETTING_WEBSITE_SITE_NAME ` is reliable moving forward for both Linux and Windows. //CC: @panchagnula 
My opinion was it is the client applications' responsibility to ensure the webapp was enabled with MSI before execute the code, so our library can make such assumption on the existence of env varaibles

Comment (yugangw-msft) [reply msrestazure/azure_active_directory.py]: @lmazuel, verify in both linux and windows webapp,  `APPSETTING_WEBSITE_SITE_NAME` does exist in both

**Comment by panchagnula** [comment]
CC:@ahmedelnably can you comment on the 'APPSETTING_WEBSITE_SITE_NAME' setting if that is a reliable setting we can use.

**Comment by ahmedelnably** [comment]
@panchagnula App Service always set the 'APPSETTING_WEBSITE_SITE_NAME' we should be able to reliably use it

**Comment by lmazuel** [comment]
Thanks for your confirmation @ahmedelnably @panchagnula !

**Pull request merged**